### PR TITLE
Add Push Notifications Capability

### PR DESCRIPTION
**Pull request by hannesmcman**
This PR is aimed at configuring the app to allow us to send remote push notifications to users using OneSignal. It may also include configuring local notifications using `react-native-push-notifications` if we want it to.

### CODE REVIEW

**Comment by hawkrives** [comment]
(For the flow types, just mark them as `: any` for now.)

File: android/app/build.gradle [hunk 1]
```diff
@@ -219,6 +221,7 @@ if (propFile.canRead()) {
 }
 
 dependencies {
+    compile project(':react-native-onesignal')
```
Comment (hawkrives) [inline android/app/build.gradle]: Sort this

File: android/app/src/main/java/com/allaboutolaf/MainApplication.java [hunk 2]
```diff
@@ -11,6 +11,7 @@
 import com.BV.LinearGradient.LinearGradientPackage;
 import com.calendarevents.CalendarEventsPackage;
 import com.facebook.react.ReactApplication;
+import com.geektime.rnonesignalandroid.ReactNativeOneSignalPackage;
```
Comment (hawkrives) [inline android/app/src/main/java/com/allaboutolaf/MainApplication.java]: Sort this

File: android/app/src/main/java/com/allaboutolaf/MainApplication.java [hunk 3]
```diff
@@ -41,6 +42,7 @@ public boolean getUseDeveloperSupport() {
     protected List<ReactPackage> getPackages() {
       return Arrays.<ReactPackage>asList(
         new MainReactPackage(),
+            new ReactNativeOneSignalPackage(),
```
Comment (hawkrives) [inline android/app/src/main/java/com/allaboutolaf/MainApplication.java]: Sort and reindent this

File: android/settings.gradle [hunk 4]
```diff
@@ -1,4 +1,6 @@
 rootProject.name = 'AllAboutOlaf'
+include ':react-native-onesignal'
+project(':react-native-onesignal').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-onesignal/android')
```
Comment (hawkrives) [inline android/settings.gradle]: Sort these

File: ios/AllAboutOlaf.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -1631,6 +1689,7 @@
 				LIBRARY_SEARCH_PATHS = (
 					"$(inherited)",
 					"\"$(SRCROOT)/$(TARGET_NAME)\"",
+					"\"$(SRCROOT)/$(TARGET_NAME)\"",
```
Comment (hawkrives) [inline ios/AllAboutOlaf.xcodeproj/project.pbxproj]: Remove dup

File: ios/AllAboutOlaf.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -1655,6 +1714,7 @@
...
```
Comment (hawkrives) [inline ios/AllAboutOlaf.xcodeproj/project.pbxproj]: Remove dup

File: ios/AllAboutOlaf/AppDelegate.m [hunk 7]
```diff
@@ -52,6 +54,9 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
   // ignore vibrate/silent switch when playing audio
   [[AVAudioSession sharedInstance] setCategory: AVAudioSessionCategoryPlayback error: nil];
 
+  self.oneSignal = [[RCTOneSignal alloc] initWithLaunchOptions:launchOptions
+                                                    appId:@"a46c6f2f-a240-4908-a359-801911e9b9ea"];
+
```
Comment (hawkrives) [inline ios/AllAboutOlaf/AppDelegate.m]: Remove extra line

File: ios/AllAboutOlaf/Info.plist [hunk 8]
```diff
@@ -87,9 +87,9 @@
 	<key>NSCalendarsUsageDescription</key>
 	<string>We use your calendar to add events to your calendar so that you remember what you wanted to attend.</string>
 	<key>NSLocationAlwaysUsageDescription</key>
-	<string>We only use your location when you would expect us to, and never when you aren&apos;t using the app.</string>
+	<string>We only use your location when you would expect us to, and never when you aren't using the app.</string>
 	<key>NSLocationWhenInUseUsageDescription</key>
-	<string>We only use your location when you would expect us to, and never when you aren&apos;t using the app.</string>
+	<string>We only use your location when you would expect us to, and never when you aren't using the app.</string>
```
Comment (hawkrives) [inline ios/AllAboutOlaf/Info.plist]: Undo

File: ios/AllAboutOlaf/Info.plist [hunk 9]
```diff
@@ -87,9 +87,9 @@
 	<key>NSCalendarsUsageDescription</key>
 	<string>We use your calendar to add events to your calendar so that you remember what you wanted to attend.</string>
 	<key>NSLocationAlwaysUsageDescription</key>
-	<string>We only use your location when you would expect us to, and never when you aren&apos;t using the app.</string>
+	<string>We only use your location when you would expect us to, and never when you aren't using the app.</string>
```
Comment (hawkrives) [inline ios/AllAboutOlaf/Info.plist]: Undo

**Review by hawkrives — DISMISSED** [review]
A bunch of little tweaks to the native project files

File: android/app/build.gradle [hunk 10]
```diff
@@ -229,6 +231,7 @@ dependencies {
     implementation project(':react-native-keychain')
     implementation project(':react-native-linear-gradient')
     implementation project(':react-native-network-info')
+    compile project(':react-native-onesignal')
```
Comment (hawkrives) [inline android/app/build.gradle]: Change this to `implementation`

File: ios/AllAboutOlaf/AppDelegate.m [hunk 11]
```diff
@@ -52,6 +54,8 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
   // ignore vibrate/silent switch when playing audio
   [[AVAudioSession sharedInstance] setCategory: AVAudioSessionCategoryPlayback error: nil];
 
+  self.oneSignal = [[RCTOneSignal alloc] initWithLaunchOptions:launchOptions
+                                                    appId:@"a46c6f2f-a240-4908-a359-801911e9b9ea"];
```
Comment (hawkrives) [inline ios/AllAboutOlaf/AppDelegate.m]: Align the `:` here, please.

```objc
self.oneSignal = [[RCTOneSignal alloc] initWithLaunchOptions:launchOptions
                                                       appId:@"a46c6f2f-a240-4908-a359-801911e9b9ea"];
```

**Review by hawkrives — CHANGES_REQUESTED** [review]
I don't think I have any more nitpicky comments after this.

**Comment by hawkrives** [comment]
@hannesmcman Did you get the extension signed?

**Comment by hannesmcman** [comment]
@hawkrives No, I think I still need you to configure that..that's why I put the `don'tmerge` label on this

**Review by hawkrives — DISMISSED** [review]

**Comment by hawkrives** [comment]
@hannesmcman Alright. I'll try to do that some time this week/end.

**Comment by hannesmcman** [comment]
@hawkrives Let me know when you think you've got that figured out, because we'll still need to add the extension to this PR before we merge.

**Comment by hawkrives** [comment]
Could you add the extension to this PR so the build can fail, so I'll know when I've got it working?

**Comment by hannesmcman** [comment]
@rye Can you test this on your android phone and make sure it works there?

**Comment by rye** [comment]
@hannesmcman sure, I'll add it to my to-do list for tonight.

**Comment by hannesmcman** [comment]
@hawkrives Is this still on your radar?

**Comment by hawkrives** [comment]
Okay, so we need a cert to sign the extension? @hannesmcman please bug me on Sunday if you haven't heard anything else by then. I'll try to do this this weekend.

**Comment by hannesmcman** [comment]
@hawkrives Yep! Will do!

**Comment by drewvolz** [comment]
@hannesmcman you mentioned this in a [comment](https://github.com/StoDevX/AAO-React-Native/pull/2566#issuecomment-386340552) above:

> we'll still need to add the extension to this PR before we merge.

Do you remember the context of the keyword _this_? Are you referring to adding the `OneSignalNotificationServiceExtension` to this PR?

File: source/app.js [hunk 12]
```diff
@@ -30,6 +31,35 @@ function getCurrentRouteName(navigationState: NavigationState): ?string {
 type Props = {}
 
 export default class App extends React.Component<Props> {
+	componentDidMount() {
+		startStatusBarColorChanger()
```
Comment (drewvolz) [inline source/app.js]: Where is the start/stop coming from? Currently it breaks things.

Comment (drewvolz) [reply source/app.js]: I feel like it might be related to the way we were telling the iOS status bar to be dark or light? @hawkrives

**Comment by hawkrives** [comment]
The last merge from master seems to have broken things.

**Comment by hannesmcman** [comment]
@drewvolz That was before I added the extension. Then I added it, which broke the build because of the lack of a proper cert.

**Comment by hawkrives** [comment]
@drewvolz this branch should run again (js-wise) as of edd0cb7

**Comment by hannesmcman** [comment]
@hawkrives how’s the cert coming? :)

**Comment by drewvolz** [comment]
Hawken passed that off to me, and I was trying to get a sense of what actually needed to be done. I haven’t been able to figure it out yet, either.

**Comment by hannesmcman** [comment]
We need to add a provisioning profile to the OneSignal Extension. Not sure if it can be the same as for the app itself.

**Comment by drewvolz** [comment]
something something

* OneSignal's tool to generate the provisioning file https://onesignal.com/provisionator
* Steps https://documentation.onesignal.com/v5.0/docs/generate-an-ios-push-certificate#section-1-provisioning

**Comment by hannesmcman** [comment]
@hawkrives I think your the only admin on the Apple developer account right? I think you're the only one able to use the tool Drew linked in the comment above...

**Comment by hawkrives** [comment]
Drew is also an admin

**Comment by hannesmcman** [comment]
@hawkrives @drewvolz Well then I guess you two will just have to arm wrestle!

**Comment by drewvolz** [comment]
@hawkrives I don't have the permissions to generate a cert on your account. I only have access to certs/identifiers/profiles from my personal developer account.

**Comment by hawkrives** [comment]
@drewvolz noted

**Review by hawkrives — APPROVED** [review]
sure

**Pull request merged**